src/Microsoft.Diagnostics.Monitoring.EventPipe/DiagnosticsMonitor.cs

```diff
@@ -0,0 +1,106 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Diagnostics.NETCore.Client;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe
+{
+    partial class DiagnosticsEventPipeProcessor
+    {
+        private sealed class DiagnosticsMonitor : IAsyncDisposable
+        {
+            private readonly MonitoringSourceConfiguration _sourceConfig;
+            private readonly TaskCompletionSource<object> _stopProcessingSource;
+            private Task _currentTask;
+
+            public DiagnosticsMonitor(MonitoringSourceConfiguration sourceConfig)
+            {
+                _sourceConfig = sourceConfig;
+                _stopProcessingSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+
+            public Task<Stream> ProcessEvents(DiagnosticsClient client, TimeSpan duration, CancellationToken cancellationToken)
```

Comment:
I've always found this parameter (`CancellationToken cancellationToken`) to be misleading on this method because it doesn't cancel the method (the one exception is the initial check in the method) and it doesn't cancel the task that this method returns. I feel like this should be removed in favor of having the caller call StopProcessing when it needs to stop processing events.

Reply:
My only concern here is whether StopSession interferes with EventPipeEventSource.StopProcessing. I can test this out, but I want to make sure we maintain the semantics of stopping without data loss.